### DISCUSSION TITLE
关于TVM Operator Inventory的文档和使用的问题

### DISCUSSION THREAD

**Discussion by Sanzo00** [mlc-ai/mlc-zh#76] [Q&A]
我在写lesson 4作业的时候，参考了TOPI的文档，遇到了两个问题：
1. **[tvm.topi.nn.pool2d](https://tvm.apache.org/docs/reference/api/python/topi.html#tvm.topi.nn.pool2d) 缺少了关于`dilation `的参数介绍**。
  <img width="681" alt="image" src="https://user-images.githubusercontent.com/36565277/181227899-9c24c853-6f9e-4e5e-ba78-220f28012866.png">
  

**2. 当我尝试使用`tvm.topi.nn.matmul`做`linear`运算的时候：**

```py
lv6 = bb.emit_te(tvm.topi.nn.matmul, lv4, linear0_weight, bias=linear0_bias, transpose_b=True) 
```
触发了一个断言错误：
```bash
assert len(bias.shape) == 1
```
我尝试去对`linear0_bias`做变换，得到如下错误：
```bash
AttributeError: <class 'tvm.relay.expr.Constant'> has no attribute reshape
```

我不太清楚`relax.cosnst`是否支持`reshape`操作。

我目前的解决方法是在变量定义的时候修改它的形状或者不使用`bias`这个参数，不知道有没有更好的解决方法。

```py
# method 1
# linear0_bias = relax.const(weight_map["linear0_bias"].reshape(1, 100), "float32")
linear0_bias = relax.const(weight_map["linear0_bias"].reshape(100), "float32")
lv6 = bb.emit_te(tvm.topi.nn.matmul, lv4, linear0_weight, bias=linear0_bias, transpose_b=True) 

# method 2
lv5 = bb.emit_te(tvm.topi.nn.matmul, lv4, linear0_weight, transpose_b=True)
lv6 = bb.emit_te(tvm.topi.add, lv5, linear0_bias)
```

**Comment by GeLee-Q** [comment]
`tvm.topi.nn.pool2d` 的dialation参数可能是  (List/ tuple of two ints)
```python
tvm.topi.nn.pool2d(data = Input, kernel = [2, 2], dilation = (1,1), stride = [2,2], padding = [0,0,0,0], pool_type = 'max')
```
我用的全连接层做的linear运算，可以作为一种思路参考, 文档里似乎并不支持bias
```
def linear_5(Input, weight, bias):
  lv5_0 = tvm.topi.nn.dense(Input, weight)
  return tvm.topi.add(lv5_0, bias) 
```

**Reply by Sanzo00** [reply]
嗯嗯，`nn.dense`和`nn.matmul`应该也支持bias，不过得是一维的。
